Improve Historic Data Retrieval of > 9 Days for `production` and `consumptionForecast` in `US-NY-NYIS` Zone

chriswasser opened the pull request:
## Issue

Currently, it is not possible to retrieve production and consumptionForecast data that is older than 9 days for the US-NY-NYIS zone:

```sh
$ poetry run test_parser US-NY-NYIS production --target_datetime 2024-01-06
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/rwthfs/rz/cluster/home/cw585968/Desktop/GitHub/electricitymaps-contrib/.venv/lib/python3.10/site-packages/click/core.py", line 1161, in __call__
    return self.main(*args, **kwargs)
  File "/rwthfs/rz/cluster/home/cw585968/Desktop/GitHub/electricitymaps-contrib/.venv/lib/python3.10/site-packages/click/core.py", line 1082, in main
    rv = self.invoke(ctx)
  File "/rwthfs/rz/cluster/home/cw585968/Desktop/GitHub/electricitymaps-contrib/.venv/lib/python3.10/site-packages/click/core.py", line 1443, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/rwthfs/rz/cluster/home/cw585968/Desktop/GitHub/electricitymaps-contrib/.venv/lib/python3.10/site-packages/click/core.py", line 788, in invoke
    return __callback(*args, **kwargs)
  File "/rwthfs/rz/cluster/home/cw585968/Desktop/GitHub/electricitymaps-contrib/test_parser.py", line 78, in test_parser
    res = parser(*args, target_datetime=parsed_target_datetime, logger=logger)
  File "/rwthfs/rz/cluster/home/cw585968/Desktop/GitHub/electricitymaps-contrib/parsers/lib/config.py", line 76, in wrapped_f
    result = f(*args, **kwargs)
  File "/rwthfs/rz/cluster/home/cw585968/Desktop/GitHub/electricitymaps-contrib/parsers/US_NY.py", line 121, in fetch_production
    raise NotImplementedError(
NotImplementedError: you can get data older than 9 days at the url http://mis.nyiso.com/public/
```

This feature limitation was already noted within the source code [here](https://github.com/electricitymaps/electricitymaps-contrib/blob/3c2cc04d59233cd4d9c55f2ea7529e2e6437431b/parsers/US_NY.py#L120-L123) and [here](https://github.com/electricitymaps/electricitymaps-contrib/blob/3c2cc04d59233cd4d9c55f2ea7529e2e6437431b/parsers/US_NY.py#L250-L255):

```python
    if (
        datetime.now(tz=TIMEZONE) - target_datetime
    ).days > 9:  # TODO: If datetime is previous to 9 days, it need to be implemented
        raise NotImplementedError(
            "you can get data older than 9 days at the url http://mis.nyiso.com/public/"
        )
```

## Description

This PR resolves these TODOs and provides the ability to retrieve historic data that lies more than 9 days in the past.

Instead of downloading an individual CSV file, it is necessary to download a ZIP file, which contains a whole month of data. From this ZIP file, a single CSV file is extracted; both these steps happen in memory to avoid temporary files. Afterwards, the usual `pd.read_csv` parsing is used to prepare the data for the same unmodified postprocessing steps. No additional dependencies are required to support this feature extension since the `zipfile` module within the Python standard library can be used. Common errors for unavailable data are caught as `RequestsHTTPError` or `KeyError` (depending on whether the error occurs during the download of the ZIP file or during the extraction of the CSV file) and an empty list is returned from the parser, adhering to the same interface used throughout the code. Otherwise, only minor typing errors were fixed within the same parser file.

### Preview

With the changes made by this commit, longer historic data can be retrieved:

```sh
$ poetry run test_parser US-NY-NYIS production --target_datetime 2024-01-06                      
Parser result:
[{'correctedModes': [],
  'datetime': datetime.datetime(2024, 1, 6, 0, 5, tzinfo=zoneinfo.ZoneInfo(key='America/New_York')),
  'production': {'gas': 7101.0, 'hydro': 2754.0, 'nuclear': 3311.0, 'unknown': 257.0, 'wind': 291.0},
  'source': 'nyiso.com',
  'sourceType': <EventSourceType.measured: 'measured'>,
  'storage': {},
  'zoneKey': 'US-NY-NYIS'},
 {'correctedModes': [],
  'datetime': datetime.datetime(2024, 1, 6, 0, 10, tzinfo=zoneinfo.ZoneInfo(key='America/New_York')),
  'production': {'gas': 7049.0, 'hydro': 2658.0, 'nuclear': 3311.0, 'unknown': 247.0, 'wind': 261.0},
  'source': 'nyiso.com',
  'sourceType': <EventSourceType.measured: 'measured'>,
  'storage': {},
  'zoneKey': 'US-NY-NYIS'},
 ...
 {'correctedModes': [],
  'datetime': datetime.datetime(2024, 1, 6, 23, 55, tzinfo=zoneinfo.ZoneInfo(key='America/New_York')),
  'production': {'gas': 7775.0, 'hydro': 2613.0, 'nuclear': 3309.0, 'unknown': 241.0, 'wind': 354.0},
  'source': 'nyiso.com',
  'sourceType': <EventSourceType.measured: 'measured'>,
  'storage': {},
  'zoneKey': 'US-NY-NYIS'},
 {'correctedModes': [],
  'datetime': datetime.datetime(2024, 1, 7, 0, 0, tzinfo=zoneinfo.ZoneInfo(key='America/New_York')),
  'production': {'gas': 7772.0, 'hydro': 2486.0, 'nuclear': 3309.0, 'unknown': 245.0, 'wind': 355.0},
  'source': 'nyiso.com',
  'sourceType': <EventSourceType.measured: 'measured'>,
  'storage': {},
  'zoneKey': 'US-NY-NYIS'}]
---------------------
took 0.91s
min returned datetime: 2024-01-06 05:05:00+00:00 UTC
max returned datetime: 2024-01-07 05:00:00+00:00 UTC 
```

### Double check

- [x] I have tested my parser changes locally with `poetry run test_parser "zone_key"`
- [x] I have run `pnpx prettier@2 --write .` and `poetry run format` in the top level directory to format my changes.

Thank you for considering this PR 🤓 Greetings

VIKTORVAV99 commented:
Looks great and works perfectly but if you could add a snapshot test that checks the output of the historical data that would be great, just to ensure there is no regressions in the future.

chriswasser commented:
Hi Viktor,

Thanks for the positive feedback 👍

Sorry, I wasn't aware of the additional parser tests in `parsers/test/` and only used `poetry run test_parser`. I just added a snapshot test for the new functionality, which queries data that lies more than 9 days in the past:

```sh
$ poetry run pytest parsers/test/test_US_NY.py
====================================================================== test session starts =======================================================================
platform linux -- Python 3.10.8, pytest-7.4.4, pluggy-1.5.0
rootdir: /rwthfs/rz/cluster/home/cw585968/Desktop/GitHub/electricitymaps-contrib
configfile: pyproject.toml
plugins: anyio-4.9.0, syrupy-4.9.0
collected 2 items                                                                                                                                                

parsers/test/test_US_NY.py ..                                                                                                                              [100%]

======================================================================== warnings summary ========================================================================
parsers/test/test_US_NY.py::test_snapshot_fetch_consumption_forecast
  /rwthfs/rz/cluster/home/cw585968/Desktop/GitHub/electricitymaps-contrib/.venv/lib/python3.10/site-packages/pandas/core/dtypes/cast.py:1641: DeprecationWarning: np.find_common_type is deprecated.  Please use `np.result_type` or `np.promote_types`.
  See https://numpy.org/devdocs/release/1.25.0-notes.html and the docs for more information.  (Deprecated NumPy 1.25)
    return np.find_common_type(types, [])

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
-------------------------------------------------------------------- snapshot report summary ---------------------------------------------------------------------
2 snapshots passed.
====================================================================== slowest 10 durations ======================================================================
1.26s setup    parsers/test/test_US_NY.py::test_snapshot_fetch_consumption_forecast
0.20s call     parsers/test/test_US_NY.py::test_snapshot_fetch_consumption_forecast
0.06s call     parsers/test/test_US_NY.py::test_snapshot_fetch_production_more_than_9_days_in_past

(3 durations < 0.005s hidden.  Use -vv to show these durations.)
================================================================== 2 passed, 1 warning in 6.21s ==================================================================
```

Since the other existing test for `US_NY.py` wasn't working due to a) using the wrong date within the `freeze_time` and b) not mocking the data retrieval correctly (as `pandas.read_csv` uses urllib in the background instead of the provided requests session), I modified the associated code as well. Now both tests run successfully 🎉

Hope everything is in order now 👌 I look forward hearing from you 🤓

Greetings

VIKTORVAV99 reviewed (approved):
Perfect!

Looks awesome and thanks for fixing the existing test as well!
